File: ModelBase/src/nodes/Node.cpp
/***********************************************************************************************************************
**
** Copyright (c) 2011, 2014 ETH Zurich
** All rights reserved.
**
** Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
** following conditions are met:
**
**    * Redistributions of source code must retain the above copyright notice, this list of conditions and the following
**      disclaimer.
**    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
**      following disclaimer in the documentation and/or other materials provided with the distribution.
**    * Neither the name of the ETH Zurich nor the names of its contributors may be used to endorse or promote products
**      derived from this software without specific prior written permission.
**
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
**
***********************************************************************************************************************/

#include "Node.h"

#include "Reference.h"
#include "UsedLibrary.h"
#include "../ModelBasePlugin.h"
#include "../ModelException.h"
#include "../commands/NodeOwningCommand.h"
#include "../commands/UndoCommand.h"
#include "../model/AllTreeManagers.h"
#include "../model/TreeManager.h"
#include "../util/ResolutionRequest.h"

#include "Core/src/AdapterManager.h"
#include "Logger/src/Log.h"

namespace Model {

DEFINE_TYPE_ID_BASE(Node, "Node", )

QHash<QString, Node::NodeConstructor> Node::nodeConstructorRegister;
QHash<QString, Node::NodePersistenceConstructor> Node::nodePersistenceConstructorRegister;

QSet<const Node*>& Node::partiallyLoadedNodes()
{
	static QSet<const Node*> set;
	return set;
}

Node::Node(Node* parent) : parent_{parent}, manager_{parent ? parent->manager_ : nullptr}
{
	if (parent && !parent->isModifyable())
		throw ModelException{"Trying to create a node with a non-modifiable parent."};
}

Node::Node(const Node&) : Node{nullptr} {}

Node::~Node()
{
	partiallyLoadedNodes().remove(this);

	for (auto m : AllTreeManagers::instance().loadedManagers())
		m->nodeIdMap().remove(this);
}

Node* Node::createDefaultInstance(Node*)
{
	Q_ASSERT(false);
	return nullptr;
}

void Node::execute(UndoCommand *command)
{
	if ( this != command->target() )
		throw ModelException{"Command target differs from current node when executing commands"};

	TreeManager* m = manager();

	bool hasOwnTreeManager = m;

	// The if below is needed to make it possible to use nodes already attached in the tree, in the set method
	// of new nodes, which are not yet attached. Without it this command will not be recorded in the undo stack.
	if (auto owning = dynamic_cast<NodeOwningCommand*>(command))
	{
		auto managerOfInserted = owning->insertedNode() ?
			AllTreeManagers::instance().managerOfOwningUndoStack(owning->insertedNode()) : nullptr;
		if (!m) m = managerOfInserted;
		else Q_ASSERT(!managerOfInserted || m == managerOfInserted);
		auto managerOfRemoved = owning->removedNode() ?
			AllTreeManagers::instance().managerOfOwningUndoStack(owning->removedNode()) : nullptr;
		if (!m) m = managerOfRemoved;
		else Q_ASSERT(!managerOfRemoved || m == managerOfRemoved);
	}


	if (m)
	{
		if (hasOwnTreeManager && !m->canBeModified(this) ) throw ModelException{"Can not modify the current node."};
		m->pushCommandOnUndoStack(command);
	}
	else
	{
		command->redo();
		SAFE_DELETE(command);
	}
}

Node* Node::lowestCommonAncestor(Node* other, Node** thisParent, Node** otherParent)
{
	QList<Node*> thisParents;
	QList<Node*> otherParents;

	// Get all parents of the current node
	Node* n = this;
	while ( n )
	{
		thisParents.prepend(n);
		n = n->parent();
	}

	// Get all parents of the other node
	n = other;
	while ( n )
	{
		otherParents.prepend(n);
		n = n->parent();
	}

	// Find the lowest common ancestor
	n = nullptr;
	while ( thisParents.size() > 0 && otherParents.size() > 0 && thisParents.first() == otherParents.first() )
	{
		n = thisParents.first();

		thisParents.removeFirst();
		otherParents.removeFirst();
	}

	Q_ASSERT(thisParents.isEmpty() == otherParents.isEmpty());
	if (thisParent)
	{
		if (thisParents.isEmpty()) *thisParent = nullptr;
		else *thisParent = thisParents.first();
	}
	if (otherParent)
	{
		if (otherParents.isEmpty()) *otherParent = nullptr;
		else *otherParent = otherParents.first();
	}

	return n;
}

bool Node::isAncestorOf(const Node* other) const
{
	if (other == nullptr) return false;

	const Node* p = other->parent();

	while (p && p != this) p = p->parent();

	return p == this;
}

Node* Node::firstAncestorOfType(const SymbolMatcher& typeMatch) const
{
	auto typeOrSubtypeMatches = [typeMatch](Model::Node* node) {
		if (typeMatch.isFixedString())
			return node->isSubtypeOf(typeMatch.matchPattern());
		return typeMatch.matches(node->typeName());
	};
	auto p = parent();
	while (p && !typeOrSubtypeMatches(p)) p = p->parent();
	return p;
}

Node* Node::childToSubnode(const Node* other) const
{
	if (other == nullptr) return nullptr;

	const Node* child = other;
	const Node* parent = child->parent();

	while (parent && parent != this)
	{
		child = parent;
		parent = child->parent();
	}

	return parent == this ? const_cast<Node*>(child) : nullptr;
}

bool Node::isModifyable() const
{
	TreeManager* m = manager();

	return !m || m->canBeModified(this);
}

bool Node::replaceChild(Node*, Node*)
{
	return false;
}

bool Node::findSymbols(std::unique_ptr<ResolutionRequest> request) const
{
	bool found{};

	if (request->direction() == SEARCH_HERE && !isTransparentForNameResolution())
	{
		if (symbolMatches(request->matcher(), request->symbolTypes()))
		{
			found = true;
			request->result().insert(const_cast<Node*>(this));
		}
	}
	else if (request->direction() == SEARCH_DOWN
				|| (request->direction() == SEARCH_HERE && isTransparentForNameResolution()))
		for (auto c : childrenInScope())
			found = c->findSymbols(request->clone(SEARCH_HERE, false)) || found;
	else if (request->direction() == SEARCH_UP || request->direction() == SEARCH_UP_ORDERED)
	{
		auto ignore = childToSubnode(request->source());
		for (auto c : childrenInScope())
		{
			// Optimize the search by skipping this scope, since we've already searched there
			if (c != ignore)
				found = c->findSymbols(request->clone(SEARCH_HERE, false) ) || found;
		}

		if ((request->exhaustAllScopes() || !found)
			 && symbolMatches(request->matcher(), request->symbolTypes()) && !isTransparentForNameResolution())
		{
			found = true;
			request->result().insert(const_cast<Node*>(this));
		}

		if ((request->exhaustAllScopes() || !found) && parent_)
			found = parent_->findSymbols(request->clone(SEARCH_UP)) || found;

		// Search in libraries. This is only valid for root nodes
		if ((request->exhaustAllScopes() || !found) && !parent_)
			for (const UsedLibrary* lib : usedLibraries())
			{
				auto libRoot = lib->libraryRoot();
				if (libRoot)
					found = libRoot->findSymbols(request->clone(libRoot, SEARCH_DOWN)) || found;
			}
	}

	return found;
}

QList<Node*> Node::childrenInScope() const
{
	return children();
}

void Node::beginModification(const QString &text)
{
	if (auto m = manager())
		m->beginModification(this, text);
}

void Node::endModification()
{
	if (auto m = manager()) m->endModification();
}

QString Node::toDebugString()
{
	auto ntdsa = Core::AdapterManager::adapt<NodeToDebugStringAdapter>(this);

	QString ret = ntdsa ? ntdsa->str : "no debug string for node";
	SAFE_DELETE(ntdsa);
	return ret;
}

bool Node::hasPartiallyLoadedChildren() const
{
	if (isPartiallyLoaded()) return true;

	return false;
}

void Node::setParent(Node* parent)
{
	parent_ = parent;
	auto oldManager = manager_;
	manager_ = parent ? parent->manager_ : nullptr;
	if (oldManager != manager_)	propagateManagerToChildren();

	if (manager_) Reference::unresolveAfterNewSubTree(this);
}

void Node::propagateManagerToChildren()
{
	QList<Node*> stack;
	stack.append(children());
	while (!stack.isEmpty())
	{
		auto top = stack.takeLast();
		top->manager_ = manager_;
		stack.append(top->children());
	}
}

void Node::setRootManager(TreeManager* manager)
{
	Q_ASSERT(!parent_);
	Q_ASSERT(manager);
	manager_ = manager;
	propagateManagerToChildren();
}

QList<Node*> Node::children() const
{
	return QList<Node*>{};
}

bool Node::definesSymbol() const
{
	return false;
}

bool Node::isTransparentForNameResolution() const
{
	return false;
}

const QString& Node::symbolName() const
{
	static QString nullString;
	return nullString;
}

Node::SymbolTypes Node::symbolType() const
{
	return UNSPECIFIED;
}

bool Node::isNewPersistenceUnit() const
{
	return false;
}

Node* Node::persistentUnitNode() const
{
	const Node* persistentUnitNode = this;
	const Node* prev = this;

	while ( persistentUnitNode && persistentUnitNode->isNewPersistenceUnit() == false )
	{
		prev = persistentUnitNode;
		persistentUnitNode = persistentUnitNode->parent();
	}

	if (persistentUnitNode) return const_cast<Node*> (persistentUnitNode);
	else return const_cast<Node*> (prev);
}

int Node::revision() const
{
	return revision_;
}

void Node::incrementRevision()
{
	revision_++;
}

void Node::addToRevision(int valueToAdd)
{
	revision_ += valueToAdd;
}

NodeReadWriteLock* Node::accessLock() const
{
	if ( parent_ ) return parent_->accessLock();
	else
		return manager()->rootLock();
}

QList<const UsedLibrary*> Node::usedLibraries() const
{
	QList<const UsedLibrary*> all;
	for (auto c : children())
		all << c->usedLibraries();
	return all;
}

void Node::registerNodeType(const QString &type, const NodeConstructor constructor,
		const NodePersistenceConstructor persistenceconstructor)
{
	if ( isTypeRegistered(type) )
		throw ModelException{"Trying to register a node type that has already been registered: " + type};

	nodeConstructorRegister.insert(type, constructor);
	nodePersistenceConstructorRegister.insert(type, persistenceconstructor);

	ModelBasePlugin::log().info("Registered new node type " + type);
}

Node* Node::createNewNode(const QString &type, Node* parent)
{
	auto iter = nodeConstructorRegister.find(type);
	if ( iter != nodeConstructorRegister.end() )
		return iter.value()(parent);
	else
	{
		ModelBasePlugin::log().error("Could not create new node. Requested node type '"
				+ type +"' has not been registered.");
		return nullptr;
	}
}

Node* Node::createNewNode(const QString &type, Node* parent, PersistentStore &store, bool partialLoadHint)
{
	auto iter = nodePersistenceConstructorRegister.find(type);
	if ( iter != nodePersistenceConstructorRegister.end() )
		return iter.value()(parent, store, partialLoadHint);
	else
	{
		ModelBasePlugin::log().error("Could not create new node from persistence. Requested node type '"
				+ type + "' has not been registered.");
		return nullptr;
	}
}

bool Node::isTypeRegistered(const QString &type)
{
	return nodeConstructorRegister.contains(type);
}

}

File: JavaImportTool/src/javaImportTool/Node.java
/***********************************************************************************************************************
 **
 ** Copyright (c) 2011, 2014 ETH Zurich
 ** All rights reserved.
 **
 ** Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
 ** following conditions are met:
 **
 **    * Redistributions of source code must retain the above copyright notice, this list of conditions and the
 **      following disclaimer.
 **    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
 **      following disclaimer in the documentation and/or other materials provided with the distribution.
 **    * Neither the name of the ETH Zurich nor the names of its contributors may be used to endorse or promote products
 **      derived from this software without specific prior written permission.
 **
 **
 ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 **
 **********************************************************************************************************************/
package javaImportTool;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintStream;
import java.io.UnsupportedEncodingException;
import java.util.LinkedList;
import java.util.List;
import java.util.ListIterator;
import java.util.Stack;

import org.apache.commons.lang3.StringEscapeUtils;

/**
 * The Node class represents an Envision model node that will be saved to the generated XML output.
 * 
 * This single class is used to model all corresponding nodes in Envision. Therefore it can represent both leaf nodes
 * and composite nodes such as lists or CompositeNode.
 */
public class Node {
	
	// These are all the properties a node might have
	private Node parent_;
	private String tag_;
	private String name_;
	private long id_;
	private String symbol_ = null; // only set this if this node defines a symbol (e.g. class or method definition)
	private String text_ = null; // only set this if this is a terminal node and contains text (e.g. a literal)
	
	// Each new node is assigned a unique id
	private static long nextId_ = 0;
	
	// These are used when writing the nodes to a stream
	private static String outputDir_ = null;
	private static Stack<PrintStream> out_ = new Stack<PrintStream>();
	private static OutputFormat format_ = OutputFormat.XML;
	
	private List<Node> children_ = new LinkedList<Node>();
	
	// Getters and setters
	public Node parent() { return parent_; }
	public String tag() { return tag_; }
	public String name() { return name_; }
	public int numChildren() { return children_.size(); }
	
	public void setName(String name) { name_ = name; }
	public void setName(int name) { name_ = Integer.toString(name); }
	public void setDoubleValue(double val) { text_ = "D_" + Double.toString(val); }
	public void setStringValue(String val) { text_ = "S_" + val; }
	public void setLongValue(long val){ text_ = "I_" + Integer.toString((int)val); }
	//TODO: In Envision, implement long integers
	
	public boolean isPersistenceUnit()
	{
		return NodeDescriptors.isPersistenceUnit(this);
	}
	
	public void setSymbol(String symbol) throws ConversionException
	{
		symbol_ = symbol;
		child("name").setStringValue(symbol);
	}
	
	// Constructors
	public Node(Node parent, String tag)
	{
		parent_ = parent;
		tag_ = tag;
		id_ = nextId_++;
		NodeDescriptors.initialize(this, tag_);
	}
	
	public Node(Node parent, String tag, String name)
	{
		this(parent, tag);
		name_ = name;
	}
	
	public Node(Node parent, String tag, int name)
	{
		this(parent, tag, Integer.toString(name));
	}
	
	// Other methods
	public Node add(Node n)
	{
		n.parent_ = this;
		children_.add(n);
		return n;
	}
	
	public Node child(String name) throws ConversionException
	{
		for (Node c : children_)
		{
			if (c.name_.equals(name)) return c;
		}
		
		throw new ConversionException("Could not find child '" + name + "' in a node of type '" + tag_ + "'");
	}
	
	public Node setChild(String name, Node newNode) throws ConversionException
	{
		ListIterator<Node> it = children_.listIterator();
		while(it.hasNext())
		{
			Node n = it.next();
			if (n.name_.equals(name))
			{
				it.set(newNode);
				newNode.parent_ = this;
				newNode.setName(name);
				return newNode;
			}
		}
		
		throw new ConversionException("Could not find child '" + name + "' in a node of type '" + tag_ + "'");
	}
	
	public Node getModuleContext(String restOfQualifiedName) throws ConversionException
	{
		return getModuleContext(restOfQualifiedName.split("\\."));
	}
	
	public Node getModuleContext(String[] restOfQualifiedName) throws ConversionException
	{
		if (restOfQualifiedName.length == 0 )
			return this;
		
		String[] remaining = new String[restOfQualifiedName.length-1];
		for (int i = 0; i <remaining.length; ++i) remaining[i] = restOfQualifiedName[i+1];
		
		for (Node c : child("modules").children_)
		{
			if (c.symbol_ != null)
			{
				if (c.symbol_.equals(restOfQualifiedName[0]))
					return c.getModuleContext(remaining);
			}
		}
		
		return addSymbolNodeInList("modules", "Module", restOfQualifiedName[0]).getModuleContext(remaining);
	}
	
	public Node addSymbolNodeInList(String listName, String tag, String symbolName) throws ConversionException
	{
		for(Node child : children_)
		{
			if (child.name_.equals(listName))
			{
				Node newNode = child.add(new Node(child, tag, child.children_.size()));
				newNode.setSymbol(symbolName);
				return newNode;
			}
		}
		
		throw new ConversionException("Symbol list '" + listName + "' is unknown in node of type '" + tag + "'");
	}
	
	public enum OutputFormat {XML, CLIPBOARD, SIMPLE};
	
	// Save as XML to a stream in either file or clipboard format
	public void renderRootTree(String dir, String projectName, OutputFormat format)
		throws ConversionException, FileNotFoundException, UnsupportedEncodingException
	{
		outputDir_ = dir;
		format_ = format;
		
		out_.push( new PrintStream(new File(outputDir_ + projectName), "UTF-8") );
		
		if (format_ == OutputFormat.XML || format_ == OutputFormat.CLIPBOARD)
			out_.peek().println("<!DOCTYPE EnvisionFilePersistence>");
		
		if (format_ == OutputFormat.CLIPBOARD) out_.peek().println("<clipboard>");
		else
		{
			setName(symbol_);
			if (format_ == OutputFormat.XML || format_ == OutputFormat.CLIPBOARD)
				out_.peek().println("<model nextid=\"" + nextId_ + "\">");
			else if (format_ == OutputFormat.SIMPLE)
				out_.peek().println("model model " + nextId_);
		}
		renderTree("\t", false);
		if (format_ == OutputFormat.XML) out_.peek().println("</model>");
		else if (format_ == OutputFormat.CLIPBOARD) out_.peek().println("</clipboard>");
	}
	
	
	private void renderTree(String indentation, boolean considerPersistenceUnits)
		throws ConversionException, FileNotFoundException, UnsupportedEncodingException
	{	
		if (considerPersistenceUnits && isPersistenceUnit() && format_ != OutputFormat.CLIPBOARD)
		{
			// Create a new file for this persistence unit
			if (format_ == OutputFormat.XML)
			{
				out_.peek().println(indentation + "<persistencenewunit name=\""
						+ name_ + "\">S_" + id_ + "</persistencenewunit>");
			}
			else if (format_ == OutputFormat.SIMPLE)
				out_.peek().println(indentation + name_ + " persistencenewunit " + id_ );
			
			out_.push( new PrintStream(new File(outputDir_ + id_), "UTF-8") );
			if (format_ == OutputFormat.XML) out_.peek().println("<!DOCTYPE EnvisionFilePersistence>");
			renderTree("", false);
			out_.pop().close();
		}
		else
		{
			if (!children_.isEmpty() && text_ != null)
				throw new ConversionException("Invalid node content. Node has both children and textual content.");
			
			// This file is not a persistence unit
			if (format_ == OutputFormat.XML || format_ == OutputFormat.CLIPBOARD)
			{
				out_.peek().print(indentation + "<" + tag_);
				if (format_ == OutputFormat.XML) out_.peek().print(" id=\"" + id_ + "\"");
				out_.peek().print(" name=\"" + name_ + "\"");
				
				if (text_ != null) out_.peek().println(">" + StringEscapeUtils.escapeXml(text_) + "</" + tag_ + ">");
				else if (children_.isEmpty()) out_.peek().println(" />");
				else
				{
					out_.peek().println(">");
					for(Node child : children_)
						child.renderTree(indentation + "\t", true);
					out_.peek().println(indentation + "</" + tag_ + ">");
				}
			}
			else if (format_ == OutputFormat.SIMPLE)
			{
				out_.peek().print(indentation + name_ + " " + tag_ + " " + id_);
				if (text_ != null) out_.peek().print(". " + escape(text_));
				out_.peek().println();
				for(Node child : children_) child.renderTree(indentation + "\t", true);
			}
		}
	}
	
	String escape(String s)
	{
		s = s.replace("\\", "\\\\");
		s = s.replace("\r", "\\r");
		s = s.replace("\n", "\\n");
		return s;
	}
}
